### PR TITLE
fix: Use `$MELTANO_ENVIRONMENT` set in `.env`

### LINKED ISSUE #7280: bug: `meltano config <plugin>` ignores `MELTANO_ENVIRONMENT` environment variable
### Meltano Version

2.15.0

### Python Version

3.11

### Bug scope

CLI (options, error messages, logging, etc.)

### Operating System

macOS Monterey 12.6

### Description

We run meltano using environment variables for as much configuration as possible. I have found what I think is a bug, in that I have the meltano environment set via an environment variable, but the `meltano config <plugin>` command ignores it.

Relevant sections of `.env`:
```shell
MELTANO_ENVIRONMENT=dev
```

and `meltano.yml`
```yaml
include_paths:
- ./environments/dev.meltano.yml
plugins:
  extractors:
  - name: tap-spreadsheets-anywhere
    variant: ets
    pip_url: git+https://github.com/ets/tap-spreadsheets-anywhere.git@34f135b45a87b468f3375c72f9d451ae6525fae7
  - name: tap-s3-mongo
    inherit_from: tap-spreadsheets-anywhere
```

and from `environments/dev.meltano.yml`
```yaml
environments:
  - name: dev
    config:
      plugins:
        loaders:
          - name: target-redshift-mongo
            config:
              schema_mapping:
                payments:
                  target_schema: payments
                profiles:
                  target_schema: customers
```

Expected behavior: I expect that running `meltano config target-redshift-mongo` here (with meltano environment set to `dev` via environment variables) includes the schema mapping defined for this plugin in the dev environment.

Observed behavior: It does not include the schema mapping. But, if I pass `--environment=dev` explicitly as a command-line argument, it does include it.

```shell
# note: MELTANO_ENVIRONMENT=dev is set in .env
$ meltano config target-redshift-mongo
2023-02-07T12:43:14.151935Z [info     ] The default environment 'dev' will be ignored for `meltano config`. To configure a specific environment, please use the option `--environment=<environment name>`.
{
  "host": "<SECRET>",
  "port": <SECRET>,
  "dbname": "<SECRET>",
  "user": "<SECRET>",
  "password": "<SECRET>",
  "s3_bucket": "<SECRET>",
  "default_target_schema": "<SECRET>",
  "copy_options": "EMPTYASNULL BLANKSASNULL TRIMBLANKS TRUNCATECOLUMNS TIMEFORMAT 'auto' COMPUPDATE OFF STATUPDATE OFF",
  "batch_size_rows": 100,
  "flush_all_streams": false,
  "parallelism": 0,
  "max_parallelism": 16,
  "disable_table_cache": false,
  "add_metadata_columns": true,
  "hard_delete": false,
  "data_flattening_max_level": 0,
  "primary_key_required": false,
  "validate_records": false,
  "skip_updates": false,
  "slices": 1
}
```

```shell
# note: MELTANO_ENVIRONMENT=dev is set in .env
$ meltano --environment=dev config target-redshift-mongo
2023-02-07T12:45:17.010451Z [info     ] Environment 'dev' is active
{
  "host": "<SECRET>",
  "port": <SECRET>,
  "dbname": "<SECRET>",
  "user": "<SECRET>",
  "password": "<SECRET>",
  "s3_bucket": "<SECRET>",
  "default_target_schema": "<SECRET>",
  "copy_options": "EMPTYASNULL BLANKSASNULL TRIMBLANKS TRUNCATECOLUMNS TIMEFORMAT 'auto' COMPUPDATE OFF STATUPDATE OFF",
  "batch_size_rows": 100,
  "flush_all_streams": false,
  "parallelism": 0,
  "max_parallelism": 16,
  "schema_mapping": {
    "payments": {
      "target_schema": "payments"
    },
    "profiles": {
      "target_schema": "customers"
    }
  },
  "disable_table_cache": false,
  "add_metadata_columns": true,
  "hard_delete": false,
  "data_flattening_max_level": 0,
  "primary_key_required": false,
  "validate_records": false,
  "skip_updates": false,
  "slices": 1
}
```

If I change the `default_environment` in `meltano.yml` (so that it's not the same as the environment I have defined in `.env`) I get the same behavior.

### Code

_No response_

### PR DESCRIPTION
**Pull request by WillDaSilva**
Context on Slack: https://meltano.slack.com/archives/C03GKHWS0HM/p1675776097120869

I decided to add some special handling for `MELTANO_ENVIRONMENT` to fix this, rather than try for some kind of generic solution for respecting env vars early on in the CLI. This was partially because we already rely on special handling for this env var and its related options/settings, so any sort of generic solution would have to content with the special handling that already exists.

I pulled out the existing logic for detecting which Meltano environment was selected into a function `detect_selected_environment`, and updated it to check `project.dotenv_env` for `MELTANO_ENVIRONMENT` as a fallback if it wasn't set terminal environment. The order of precedence is:
1. `$MELTANO_ENVIRONMENT` set in the terminal
2. `$MELTANO_ENVIRONMENT` set in `.env`
3. The `--environment` CLI option.

@tayloramurphy Please let me know if that precedence is correct.

Closes #7280

### CODE REVIEW

**Comment by tayloramurphy** [comment]
@WillDaSilva I'm referencing my comment in https://github.com/meltano/meltano/issues/6386#issuecomment-1195887590 when I talk about this. From there we laid out:

```
- environment level plugin env # highest
- schedule level env
- job level env
- root level plugin env 
- environment level env 
- root level env
- .env file
- terminal env # lowest
```

```
command line flag # highest
environment level plugin env
... same as above ...
.env file
terminal  # lowest
```

My logic being that if you're setting something at run time in the command, you want that to take precedence over everything. @aaronsteers will you validate my logic and structure here please?

**Comment by WillDaSilva** [comment]
@tayloramurphy We previously used [Click](https://click.palletsprojects.com/en/8.1.x/options/#values-from-environment-variables) for this, and Click makes the env var take precedence over the CLI option. We can change the precedence per your request, but we should be aware that that would make this a change to the existing behaviour, rather than a mere addition.

**Comment by tayloramurphy** [comment]
@WillDaSilva ok, good to know. Let's get @aaronsteers's buy-in on this. It seems like if we have a current behavior we need to validate it's documented properly in https://docs.meltano.com/guide/configuration#specifying-environment-variables and then flag this sort of change for 3.0 since it's breaking.

**Comment by aaronsteers** [comment]
@tayloramurphy and @WillDaSilva - I always defer to [this site](https://clig.dev/#configuration) for reference:

> image

I agree with this order, and I think if a user specifally sets `--environment=prod`, that should override the terminal context and `.env` file.

We should probably get this into our own handbook too.

**Comment by tayloramurphy** [comment]
@aaronsteers to be clear then, what you're proposing based on that site is:

```
- command line flag # highest
- environment level plugin env
- schedule level env
- job level env
- root level plugin env 
- environment level env 
- root level env
- terminal 
- .env file  # lowest
```

I'm fine with that as well - I just want us to commit to something clearly and then make progress on making our stated target reality.

**Comment by WillDaSilva** [comment]
@tayloramurphy @aaronsteers @edgarrmondragon I've updated the precedence and added a test for it here: ebdb940ddecf34254c02b24573cfab165a88f971

**Comment by aaronsteers** [comment]
> @tayloramurphy and @WillDaSilva - I always defer to [this site](https://clig.dev/#configuration) for reference:
> 
> > image

@WillDaSilva - Side-by-side, your code in https://github.com/meltano/meltano/commit/ebdb940ddecf34254c02b24573cfab165a88f971 comment matches the above perfectly, I think:

> Precedence is:
> 1. The `--environment` CLI option
> 2. Env var from the shell
> 3. Env var from `.env`
> 4. Default environment from `meltano.yml`
> 5. `None`

----------

> @aaronsteers to be clear then, what you're proposing based on that site is:
> 
> ```
> - command line flag # highest
> - environment level plugin env
> - schedule level env
> - job level env
> - root level plugin env 
> - environment level env 
> - root level env
> - terminal 
> - .env file  # lowest
> ```

I need to double-check some of these before I can say yes to the above. For this PR though, we are just handling the "--environment" CLI arg and `MELTANO_ENVIRONMENT`, which I think are not applicable at the other levels.

**Comment by tayloramurphy** [comment]
@aaronsteers happy to spin it out. We've talked a lot about this and I feel like we keep losing our previous decisions on what the target state is supposed to be. This issue:

* https://github.com/meltano/meltano/issues/6386

Has a Miro board along with several threads where we hashed out the order of precedence. I will need help to get everyone on the same page with this cc @seth-meltano 

----

For this PR though I think the proposed precedence looks good and I'll approve.

**Review by tayloramurphy — APPROVED** [review]

File: docs/src/_getting-started/index.md [hunk 1]
```diff
@@ -127,7 +127,7 @@ This will allow you to use [`git diff`](https://git-scm.com/docs/git-diff) to ea
 
 ## View and Activate Your Environments
 
-As part of creating your Meltano project, we automatically added your first [environments](/concepts/environments) called `dev`, `staging` and `prod`. This allows you to define configurations specific to the environment in which you're running your project. Theres also a [`default_environment`](https://docs.meltano.com/concepts/environments#default-environments) setting in the `meltano.yml` that get automatically set to `dev`, you can list and change the active environment using:
+As part of creating your Meltano project, we automatically added your first [environments](/concepts/environments) called `dev`, `staging` and `prod`. This allows you to define configurations specific to the environment in which you're running your project. There's also a [`default_environment`](https://docs.meltano.com/concepts/environments#default-environments) setting in the `meltano.yml` that get automatically set to `dev`, you can list and change the active environment using:
```
Comment (aaronsteers) [inline docs/src/_getting-started/index.md:130]: Do I understand correctly that this PR doesn't impact how `default_environment` does or does not get applied? For instance, `default_environment` would today be ignored for `meltano config` but respected for `meltano run`.

My observation is that this part looks unchanged in the PR (which is a good thing for preserving those behaviors) and I just want to confirm that expectation is correct.

@WillDaSilva - Do I understand that correctly?

Comment (WillDaSilva) [reply docs/src/_getting-started/index.md:130]: @aaronsteers That is correct. This PR has no impact on the application of `default_environment`.

Comment (aaronsteers) [reply docs/src/_getting-started/index.md:130]: Great. Thanks for confirming.

**Review by aaronsteers — APPROVED** [review]
🙌

**Pull request merged**